cmd/flat-network-controller/app/controller.go

```diff
@@ -0,0 +1,247 @@
+	netClientset *arktos.Clientset
+	svcClientset *kubernetes.Clientset
+	recorder     record.EventRecorder
+}
+
+// New creates the controller object
+func New(netClientset *arktos.Clientset, svcClientset *kubernetes.Clientset, informer arktosinformer.NetworkInformer) *Controller {
+	utilruntime.Must(arktoscheme.AddToScheme(scheme.Scheme))
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(klog.Infof)
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: svcClientset.CoreV1().EventsWithMultiTenancy(metav1.NamespaceAll, metav1.TenantAll)})
+
+	return &Controller{
+		store:        informer.Lister(),
+		queue:        workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		cacheSynced:  informer.Informer().HasSynced,
+		netClientset: netClientset,
+		svcClientset: svcClientset,
+		recorder: eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "flat-network-controller"}),
+	}
+}
+
+// Run starts the control loop with workers processing the items
+func (c *Controller) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Info("starting flat network controller")
+	klog.V(5).Info("waiting for informer caches to sync")
+	if !cache.WaitForCacheSync(stopCh, c.cacheSynced) {
+		klog.Error("failed to wait for cache to sync")
+		return
+	}
+
+	klog.V(5).Info("staring workers of flat network controller")
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	klog.V(5).Infof("%d workers started", workers)
+	<-stopCh
+	klog.Info("shutting down flat network controller")
+}
+
+func (c *Controller) runWorker() {
+	for {
+		item, shutdown := c.queue.Get()
+		if shutdown {
+			break
+		}
+
+		c.process(item)
+	}
+}
+
+// processItem will read a single work item off the work queue and attempt to process it
+func (c *Controller) process(item interface{}) {
+	defer c.queue.Done(item)
+
+	key, ok := item.(string)
+	if !ok {
+		klog.Errorf("unexpected item in queue: %v", item)
+		c.queue.Forget(item)
+		return
+	}
+
+	tenant, name, err := parseKey(key)
+	if err != nil {
+		klog.Errorf("unexpected string in queue; discarding: %s", key)
+		c.queue.Forget(item)
+		return
+	}
+
+	net, err := c.store.NetworksWithMultiTenancy(tenant).Get(name)
+	if err != nil {
+		klog.Warningf("failed to retrieve network in local cache by tenant %s, name %s: %v", tenant, name, err)
+		c.queue.Forget(item)
+		return
+	}
+
+	if net.Spec.Type != "flat" {
+		klog.V(5).Infof("network %s/%s is of type %q; ignored", net.Tenant, net.Name, net.Spec.Type)
+		c.queue.Forget(item)
+		return
+	}
+
+	klog.V(5).Infof("processing network %s/%s", net.Tenant, net.Name)
+
+	if err := manageFlatNetwork(net, c.netClientset, c.svcClientset); err != nil {
+		c.queue.AddRateLimited(key)
+		c.recorder.Eventf(net, corev1.EventTypeWarning, "FailedProvision", "failed to provision network %s/%s/%s: %v", net.Tenant, net.Tenant, net.Name, err)
```

Comment:
"net.Tenant, net.Tenant" -> should be  "net.Tenant, net.Namespace"

Reply:
The item is not "forget" if the network provision fails. Is it intentional? The logic here means the system will keep on trying provisioning without a limit.

Reply:
1. network object has no ns; will update the log message accordingly.
2. yes it is intentional for controller to have the loop trying to provision network eventually.